src/vdpvis.1
.TH "vdpvis" 1
.SH NAME
vdpvis \- Visualize output from vdpcmp command
.SH SYNOPSIS
.B vdpvis
.SH DESCRIPTION
Use this command to visualize the output from the Visual Difference
Predicator (vdpcmp command). The command is normally invoked
automatically form the \fIvdp\fR script.
.P
The generated images use the following color-coding to describe the
probability of detection:
.TP 14
<25% 
- gray
.TP
50% 
- green
.TP
62.5% 
- yellow
.TP
75% to 95%
 - red 
.TP
100% 
- magenta
.SH OPTIONS
.TP
.B \--sumary <file.cvs>, -s <file.cvs>
Write summary information to <file.cvs>. Summary information contains
the ratios of pixels that contain distortion of the probability
greater than 75% (first number) and 95% (second number). The numbers
are separated by a comma. Value 0 denotes no pixels, 1 all pixels.
.SH "SEE ALSO"
.BR vdp (1)
.BR vdpcmp (1)
.BR pfsin (1)
.BR pfsout (1)
.SH BUGS
Not known.


ChangeLog
        * Fixed:  Configure script handles -enable-debug with overwriting
	CXXFLAGS 

	* Added: New option: pixels-per-degree
	
v1.7.1 2010-12-19  Rafal Mantiuk  <mantiuk@gmail.com>

        * Added: Documented color coding in the manual of vdpvis
	
	* Added: bash auto-detection in configure.ac
	
	* Fixed: pupil diameter formula (previous fix was wrong)
	
	* Fixed: vdp script recognizes BMP files and is run with bash instead of sh

	* Fixed: pupil diameter replaced with pupil radius in Merimont &
	Wandel's OTF (thanks to Kenichiro Masaoka)

v1.7 2008-08-21  Rafal Mantiuk  <mantiuk@mpi-inf.mpg.de>

	* Fixed: spurious zeros in the fan filters (thanks to Kenichiro Masaoka)
	
	* Fixed: frequency misalignment bug (thanks to Kenichiro Masaoka)

	* Fixed: pupil diameter formula 
	
	* Added: support for multithreaded FFTW (though I could not observe any speed up)	
	
	* Added: --sumary option to vdpvis (dump percent summary to a text
	file)

	* Fixed: target and mask image order was wrong in some programs

	* Fixed: if OTF is used, NTF (neural) is used instead of CSF
		
v1.6 2006-06-13  Rafal Mantiuk  <mantiuk@mpi-inf.mpg.de>

	* Fixed: previous version could return different results on
	differents runs because fftw library used different IFFT code,
	although the difference was minimal. To make VDP results
	deterministic, FFTW optimizations are switched to FFTW_ESTIMATE.	

v1.5 2006-02-21  Rafal Mantiuk  <mantiuk@mpi-inf.mpg.de>
	
	* Added: ABS fix, which compares target and mask and zeros
	predictions if the pixels are almost identical

	
v1.4 2005-12-02  Rafal Mantiuk  <mantiuk@mpi-inf.mpg.de>

	* Added: Remove Inf, Nan and nonpositive values from luminance
	channels before any processing

v1.3 2005-05-09  Rafal Mantiuk  <mantiuk@mpi-inf.mpg.de>
	
	* Low dynamic range images are handled automatically by vdp script

	* Fixed to work with pfstools 1.2
	
	* Added switch --multiply-lum
	
	* Added more comments and copyright information

v1.2 2004-11-22  Rafal Mantiuk  <mantiuk@mpi-inf.mpg.de>
	
	* Added new OTFs: Marimont, Deeley (suggested)
	
	* Added option to the vdp script: -o <raw_vdp_output.pfs>
	
	* The default parameters set to the results of calibration
	procedure from SPIE'05 paper

v1.1 2004-09-23  Rafal Mantiuk  <mantiuk@mpi-inf.mpg.de>
	 
	* Added: phase uncertainty. Now the VDP works twice as slower as
	it requires additional filtering step for each band. However this
	should predict masking more acuratelly.
	
	* Fixed: Original VDP parameters set to Daly's settings
	
	* Fixed: Peak sensitivity of HDR VDP changed to 0.005% contrast,
	to match the original VDP
	
	* Added: possibility to specify viewing conditions (options or
	config file)
	 
v1.0 2004-09-21  Rafal Mantiuk  <mantiuk@mpi-inf.mpg.de>

	* Initial release

	  


src/vdpvis.cpp
/**
 * @brief visualization of the HDR VDP result 
 * 
 * This file is a part of HDR VDP.
 * ---------------------------------------------------------------------- 
 * Copyright (C) 2003,2004 Rafal Mantiuk and Grzegorz Krawczyk
 * 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * ---------------------------------------------------------------------- 
 *
 * @author Rafal Mantiuk, <mantiuk@mpi-inf.mpg.de>
 *
 * $Id: vdpvis.cpp,v 1.2 2008/08/21 17:48:00 rafm Exp $
 */

#include <iostream>
#include <fstream>
#include <stdlib.h>
#include <stdio.h>
#include <math.h>

#include <pfs.h>
#include <getopt.h>

#define FIFTY_THRESHOLD       0.50
#define FIFTY1_THRESHOLD      0.625
#define PROBABILITY_THRESHOLD 0.75
#define ALMOST_SURE_THRESHOLD 0.95


#define PROG_NAME "vdpvis"

void colorCode( const pfs::Array2D *vdp, const pfs::Array2D *L, pfs::Array2D *R, pfs::Array2D *G, pfs::Array2D *B );

void computePercentageOfPixels( const pfs::Array2D *vdp, const float prob1, float &procPixelsProb1,
  const float prob2, float &procPixelsProb2 );


class QuietException 
{
};

static void errorCheck( bool condition, const char *string )
{
  if( !condition ) {
    throw pfs::Exception( string );
  }
}

void printHelp()
{
  fprintf( stderr, PROG_NAME " [--summary=file.csv] [--verbose] [--help]\n"
    "See man page for more information.\n" );
}

void visualizeVDP(int argc, char **argv)
{
  bool verbose = false;
  const char *outputFile = NULL;

  static struct option cmdLineOptions[] = {
    { "help", no_argument, NULL, 'h' },
    { "verbose", no_argument, NULL, 'v' },
    { "summary", required_argument, NULL, 's' },
    { NULL, 0, NULL, 0 }
  };

  int optionIndex = 0;
  while( 1 ) {
    int c = getopt_long(argc, argv, "s:hv", cmdLineOptions, &optionIndex);
    if( c == -1 ) break;
    switch( c ) {
    case 'h':
      printHelp();
      throw QuietException();
    case 'v':
      verbose = true;
      break;
    case 's':
      outputFile = optarg;      
      break;      
    case '?':
      throw QuietException();
    case ':':
      throw QuietException();
    }
  }


  pfs::DOMIO pfsio;
    
  // Load frame with VDP output from stdin
  pfs::Frame *frame = pfsio.readFrame( stdin );
  errorCheck( frame != NULL, "Frame not found in pfs stream" );
  pfs::Array2D *vdp = frame->getChannel( "VDP" );
  errorCheck( vdp != NULL, "There is no vdp output in the input frame" );
  
  pfs::Channel *Y = frame->getChannel( "Y" );
  errorCheck( Y != NULL, "There is no luminance channel in the input frame" );

  float percent75, percent95;
  computePercentageOfPixels( vdp, 0.75f, percent75, 0.95f, percent95 );
  fprintf( stderr, "Probability of detection P>75%% for %2.4f%% of pixels\n", percent75*100.f );
  fprintf( stderr, "Probability of detection P>95%% for %2.4f%% of pixels\n", percent95*100.f );
  if( outputFile != NULL ) {
    std::ofstream file( outputFile );
    file << percent75 << ", " << percent95 << "\n";
    file.close();    
  }
  
  pfs::Channel *L = Y;          // Lightness
  {                             // Tone map image and compute lightness
    const int size = Y->getCols()*Y->getRows();

    // TODO: Take also variance into account to modify 0.63
    float count = 0;
    float logMean = 0;    
    for( int i = 0; i < size; i++ )
      if( (*Y)(i) > 0.0001 ) { 
        logMean += log( (*Y)(i) );
        count++;
      }
    logMean = exp( logMean / count );    
    
    for( int i = 0; i < size; i++ ) {
      (*L)(i) = 1 - 1/(1+powf( (*Y)(i)/logMean, 0.63 ));
    }
  }
  

  pfs::Channel *X, *Z;
  X = frame->createChannel( "X" );
  Z = frame->createChannel( "Z" );
  
  pfs::Channel *R = X, *G = Y, *B = Z;

  colorCode( vdp, L, R, G, B );
  
  pfs::transformColorSpace( pfs::CS_RGB, R, G, B, pfs::CS_XYZ, X, Y, Z );

  frame->getTags()->setString( "LUMINANCE", "DISPLAY" ); // LDR image
  
  pfsio.writeFrame( frame, stdout );

  pfsio.freeFrame( frame );

  
  
}


int main( int argc, char* argv[] )
{
  try {
    visualizeVDP( argc, argv );
  }
  catch( pfs::Exception ex ) {
    fprintf( stderr, PROG_NAME " error: %s\n", ex.getMessage() );
    return EXIT_FAILURE;
  }        
  catch( QuietException  ex ) {
    return EXIT_FAILURE;
  }        
  return EXIT_SUCCESS;
}

/**
 * Count pixels that have higher probability than prob1 and prob2
 */

void computePercentageOfPixels( const pfs::Array2D *vdp, const float prob1, float &procPixelsProb1, const float prob2, float &procPixelsProb2 )
{
  const int size = vdp->getCols()*vdp->getRows();

  procPixelsProb1 = 0;
  procPixelsProb2 = 0;
  
  for( int i = 0; i < size; i++ ) {
    const float prob = fabs( (*vdp)(i) );
    if( prob >= prob1 ) procPixelsProb1++;
    if( prob >= prob2 ) procPixelsProb2++;
  }

  procPixelsProb1 /= (float)size;
  procPixelsProb2 /= (float)size;
}


void colorCode( const pfs::Array2D *vdp, const pfs::Array2D *L, pfs::Array2D *R, pfs::Array2D *G, pfs::Array2D *B )
{
  const int size = vdp->getCols()*vdp->getRows();
  
  // Color coding
  for( int i = 0; i < size; i++ ) {
    const float prob = fabs( (*vdp)(i) );
    const float l = (*L)(i) * 0.7 + 0.2;    // Lightness (scaled to
                                            // avoid desaturated
                                            // colors for low and high
                                            // lightness

      // My simple coloring
//     if( prob < FIFTY_THRESHOLD ) {
//       (*R)(i) = 0.f;
//       (*G)(i) = 1.f;
//       (*B)(i) = 0.f;
//     } else if( prob < PROBABILITY_THRESHOLD ) {
//       (*R)(i) = 1.f;
//       (*G)(i) = 1.f;
//       (*B)(i) = 0.f;
//     } else if( prob < ALMOST_SURE_THRESHOLD ) {
//       (*R)(i) = 1.f;
//       (*G)(i) = 0.f;
//       (*B)(i) = 0.f;
//     } else {
//       (*R)(i) = 1.f;
//       (*G)(i) = 0.5f;
//       (*B)(i) = 0.5f;
//     }


      // Karol's PAPUGA coloring
      if( prob < PROBABILITY_THRESHOLD ) {
        if( prob > FIFTY_THRESHOLD ) {
          if( prob > FIFTY1_THRESHOLD ) {
            float t = (prob - FIFTY1_THRESHOLD) / (0.5 * (PROBABILITY_THRESHOLD - FIFTY_THRESHOLD));
            (*R)(i) = 1.0;
            (*G)(i) = 1.0 - t;
            (*B)(i) = 0.0;
          } else {
            float t = (prob - FIFTY_THRESHOLD) / (0.5 * (PROBABILITY_THRESHOLD - FIFTY_THRESHOLD));
            (*R)(i) = t;
            (*G)(i) = 1.0;
            (*B)(i) = 0.0;
          }
        } else {
          if( prob < FIFTY_THRESHOLD*0.5 ) {
            (*R)(i) = 0.8f;
            (*G)(i) = 0.8f;
            (*B)(i) = 0.8f;
          } else {
            float t = prob / FIFTY_THRESHOLD;
            (*R)(i) = 0.0;
            (*G)(i) = t;
            (*B)(i) = 0.0;
          }
        }
      } else if( prob > ALMOST_SURE_THRESHOLD ) {
        float t = (prob - ALMOST_SURE_THRESHOLD) / (1.0 - ALMOST_SURE_THRESHOLD);
        (*R)(i) = 1.f;
        (*G)(i) = 0.f;
        (*B)(i) = 0.5f * t;
      } else {
        (*R)(i) = 1.f;
        (*G)(i) = 0.f;
        (*B)(i) = 0.f;
      }    

      // Karol's no PAPUGA coloring        
//     if( prob < PROBABILITY_THRESHOLD )
//     {
//       float t = prob / PROBABILITY_THRESHOLD;
//       t -= 0.5;
//       (*R)(i) = 3.f * t;
//       (*G)(i) = 3.f * t;
//       (*B)(i) = 0.f;
//     }
//     else if( prob > ALMOST_SURE_THRESHOLD ) {
//       float t = (prob - ALMOST_SURE_THRESHOLD) / (1.0 - ALMOST_SURE_THRESHOLD);
//       (*R)(i) = 1.f;
//       (*G)(i) = 0.f;
//       (*B)(i) = 0.5f * t;
//     } else {
//       (*R)(i) = 1.f;
//       (*G)(i) = 0.f;
//       (*B)(i) = 0.f;
//     }    

      // Influence of lightness
       (*R)(i) *= l;
       (*G)(i) *= l;
       (*B)(i) *= l;
    
    
  }
}
